File: Ywy/JavaScript/level2/ParenthesesConversion.js
export default class ParenthesesConversion {
    solution = (p) => {
        if (p.length === 2) {
            return "()";
        } else if (p.length === 0) {
            return "";
        }

        const answer = [];
        const obj = this.search([...p]);
        let str = obj.u.join("");
        p = p.replace(str, "");

        if (obj.isBalance) {
            answer.push(str);
            answer.push(this.solution(p));
        } else {
            str = str.substring(1, str.length - 1);
            answer.push("(");
            answer.push(this.solution(p));
            answer.push(")");
            if (str.length > 0) {
                answer.push(this.reversal([...str]));
            }
        }
        return answer.join("");
    };

    search = (parentheses) => {
        const result = {};
        result.u = [];
        let leftCount = 0;
        let rightCount = 0;
        let index = 0;
        let range = parentheses.length - 1;
        let current = undefined;
        let next = undefined;
        while (index < range) {
            current = parentheses[index++];
            next = parentheses[index++];
            if (current === "(") {
                leftCount++;
            } else if (current === ")") {
                rightCount++;
            }
            if (next === "(") {
                leftCount++;
            } else if (next === ")") {
                rightCount++;
            }
            result.u.push(current);
            result.u.push(next);
            if (leftCount === rightCount) {
                break;
            }
        }
        return this.balanceCheck(result);
    };

    balanceCheck = (result) => {
        let index = 0;
        let range = result.u.length;
        let stack = [result.u[index]];
        let next = undefined;
        let current = result.u[index];
        while (++index < range) {
            next = result.u[index];
            if (current === "(" && next === ")") {
                current = stack.pop();
                continue;
            }
            stack.push(next);
            current = next;
        }
        result.isBalance = stack.length === 0;
        return result;
    };

    reversal = (parentheses) => {
        let index = 0;
        let range = parentheses.length;
        let at = undefined;
        while (index < range) {
            at = parentheses[index];
            parentheses[index++] = at === "(" ? ")" : "(";
        }
        return parentheses.join("");
    };
}

File: Ywy/JAVA/src/level2/ParenthesesConversion.java
package level2;

public class ParenthesesConversion {
	private boolean isBalance = true;
	public String solution(String p) {
		if (p.length() == 2) {
			return "()";
		} else if (p.isEmpty()) {
			return "";
		}
		final StringBuffer sb = new StringBuffer();
		String str = null;
		while (true) {
			str = search(p);
			p = p.substring(str.length());
			if (isBalance) {
				sb.append(str);
				sb.append(solution(p));
				return sb.toString();
			}
			str = str.substring(1, str.length() - 1);
			sb.append("(");
			sb.append(solution(p));
			sb.append(")");
			sb.append(reversal(str));
			return sb.toString();
		}
	}
	
	private String search(String p) {
		isBalance = false;
		char[] parentheses = p.toCharArray();
		int index = 0;
		int range = parentheses.length - 1;
		int leftCount = 0;
		int rightCount = 0;
		char current = ' ';
		char next = ' ';
		do { 
			current = parentheses[index++];
			next = parentheses[index];
			
			if (current == '(') {
				leftCount++;
			} else if (current == ')') {
				rightCount++;
			}
			
			if (next == '(') {
				leftCount++;
			} else if (next == ')') {
				rightCount++;
			}
			
			if (rightCount == leftCount) {
				index = rightCount + leftCount;
				break;
			}
		} while(++index < range);
		range = index;
		index = 0;
		char[] result = new char[range];
		char[] stack = new char[range];
		int top = 0;
		current = ' ';
		while (index < range) {
			next = parentheses[index];
			result[index++] = next;
			if (current == '(' && next == ')') {
				top += top > 0 ? -1 : 0;
				current = stack[top];
				continue;
			}
			stack[top++] = next;
			current = next;
		}
		
		if (top == 0) {
			isBalance = true;
		}
		return new String(result);
	}
	
	private String reversal(String p) {
		char[] parentheses = p.toCharArray();
		int index = 0;
		int range = p.length();
		char at = ' ';
		while (index < range) {
			at = parentheses[index];
			parentheses[index++] =  at == '(' ? ')' : '(';
		}
		return new String(parentheses);
	}
}
